src/composer.ts

```diff
@@ -34,6 +34,35 @@ function getText(
   return undefined
 }
 
+type Props = {
+  [key in tt.UpdateType]: tt.UpdateProps[key] & tt.ContextProps[key]
+}
+type SubProps = {
+  [key in tt.MessageSubType]: Props['message'] &
+    tt.UpdateSubProps[key] &
+    tt.ContextSubProps[key]
+}
+
+type MatchedContext<
+  C extends Context,
+  T extends tt.UpdateType | tt.MessageSubType
+> = C &
+  // TODO: support subtypes for: callback query, channel post, maybe more?
+  (T extends tt.UpdateType ? Props[T] : SubProps[Exclude<T, tt.UpdateType>]) &
+  tt.AbsentProps<
+    Exclude<
+      tt.UpdateType,
+      T | (T extends tt.MessageSubType ? 'message' : never)
+    >
+  >
+
+type EntityMatch<
```

Comment:
Confusing name, sounds like it's related to `tt.MessageEntity`. Best I can come up with is `MatchedMiddlewares`, but I think just inlining the type would be the clearest.

Reply:
Well it's called like [that](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L173) [because](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L178) [it](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L184) [has](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L191) [everything](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L198) [to](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L205) [do](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L212) [with](https://github.com/telegraf/telegraf/blob/3d8d132e658e7bfb98ae8a269b76faa39fc2dac9/src/composer.ts#L219) `tt.MessageEntity`. If taking in regular expressions makes the name less fitting, we can inline it in that one case. I don't see the need for repeating the `match` property countless times, that's a bit too WET

Reply:
Objection! You used `EntityMatch` in `hears` as well, which has nothing to do with `tt.MessageEntity`.

Reply:
That's what I was referring to with regular expressions

Reply:
As in, we can inline it for `hears` only.

Or, of course, find a better name.

Reply:
Anything wrong with `MatchedMiddlewares`? Except, of course, library users having no clue what it is, but that'd be a problem with every type alias.